keps/sig-release/20200122-kubernetes-yearly-support-period.md

```diff
@@ -0,0 +1,322 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+The goal of this KEP is to increase patch-fix support for each Kubernetes release from the current 9 months to 1 year.
+The primary motivation is to address an end-user request:  Business cycles (financial, certification, sales, vacation, HR and others) are overwhelmingly annual, leading to many users desiring at least annual support cycles in projects on which they depend.
+
+Details of the expected tasks involved in implementing this policy are outlined below, the most major of which are:
+
+- Review skew policies, identify which skews are derived from the number of supported versions, and identify new target skews (e.g. oldest kubelet against newest API server)
+- Review deprecation policies, identify which deprecation periods are derived from the number of supported versions, and identify new target deprecation periods (e.g. deprecation period for GA features/APIs)
+- Review support policies of external dependencies and identify/resolve gaps to ensure we do not claim longer support than the things we build on top of
+
+## Motivation
+
+Today the Kubernetes project delivers minor releases (e.g.: 1.13 or 1.14) every 3 months.
+The project provides bugfix support via patch releases (e.g.: 1.13.Y), with each minor release (e.g.: 1.13) having this patch release stream of support for approximately 9 months.
+This means a cluster operator must upgrade at least every 9 months to remain supported.
+Nine months is an irregular period of time relative to the planning, operations, and financial cycles of most companies.
+
+Beyond availability of patch support and an upgrade path, end-user also faces the complexity of managing usage of features versus [deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/) and the [version skew policy](https://kubernetes.io/docs/setup/release/version-skew-policy/).
+With a new minor release coming every three months, there is a constant need to evaluate and plan and enact forward movement.
+
+The survey conducted in early 2019 by the WG LTS showed that a significant subset of Kubernetes end-users fail to upgrade within the 9-month support period.
+In the graph below, 1.13 was the current version, meaning that 1.9 and 1.10 were a few months out of support.
+Yet about 1/3 of users had production machines on those versions:
+
+![Kubernetes versions in production](./20200122-versions-in-production.png)
+
+This, and other responses from the survey, suggest that this 30% of users would better be able to keep their deployments on supported versions if the patch support period were extended to 12-14 months.
```

Comment:
We did a lot of thinking, discussing, researching on this, but it's fallible and fuzzy.  We know:

 1. Some will never upgrade (can't fix)
 1. Some upgrade only when they must to remain in support (this KEP enables them to delay)
 1. Some businesses do annual upgrades (this KEP helps them): we have concrete reports from end users that this is how they operate and it's not unbelievable that this might be a thing in industry
 1. Some upgrade faster (this KEP doesn't impact them)

As a project we don't have strong incentives in place to get people in bucket 2 to move instead to buckets 3 or 4.  We already see people running releases outside of the community support lifetime.  The weighting of users running older releases may be increased because of this KEP.  Regardless of this KEP, my suspicion is the bigger nudge in that direction will be the project becoming more production ready and having more stable/ga v1 APIs.

In this balance, we have undeniable that as a project we have limited resources.  There's clear resistance from our developer base to not go to longer and longer support lifetimes, even if users ask for longer and longer, while also a recognition that annual is not unrealistic to be asked for and to consider delivering.  The gap is something vendors can and will address if there is value in addressing it.

The underlying question is for how long should a project like ours offer patch release support, component interoperability, and upgrade-ability?

Reply:
Also linked in "Alternatives" is a @timothysc discussion document around "bi-modal" or "tick-tock" with monthly/annual modes.

Our thinking was just as our current three month cycle is relatively arbitrary, shifting to monthly dev cycles plus the second mode of annual stable cycles is arbitrary.  That's why we left that for separate discussion.  Regardless of how the project chooses to deliver stable features (monthly, quarterly, three times a year, annually?), how do end users receive that and how long can they run it once deployed?  There's no one right answer.

Reply:
Two things come to mind on this...

1. Many people use the version provided by their public cloud. Those versions are typically keeping at N-2 (the last supported version in the current scheme). They also default to one version older than that being the latest version not supported. If people follow defaults or upgrade when the public cloud providers provide it they will often be out of support or just barely in it. **We should understand why cloud providers are so slow to provider newer versions.**
2. There are companies that have certain cycles. For example, I don't expect many shopping businesses to update in the shopping season in the fall. There are other companies that have to deal with tax season. Companies have cycles and a year appeared to be a good support window. On Helm we looked at something shorter than a year for Helm v2 when v3 came out. We were asked to extend it to handle businesses longer planning and action cycles. **If we hold another survey of operators it might be worth asking about their cycles for upgrades and changes.**